### PR TITLE
Refactor association traits

### PR DESCRIPTION
**Pull request by lorenzo**
This moves the logic in `SelectableAssociationTrait` and `ExternalAssociationTrait` into separate classes that are used as collaborators in the association classes.

I feel this makes the code more readable and easier to grep, as from previous experience, jumping in between the traits was quite a bad experience.

### CODE REVIEW

File: src/ORM/Association.php [hunk 1]
```diff
@@ -751,6 +751,20 @@ public function deleteAll($conditions)
...
     /**
+     * Returns true if the eager loading process will require a set of the owning table's
+     * binding keys in order to use them as a filter in the finder query.
+     *
+     * @param array $options The options containing the strategy to be used.
+     * @return bool true if a list of keys will be required
+     */
+    public function requiresKeys(array $options = [])
+    {
+        $strategy = isset($options['strategy']) ? $options['strategy'] : $this->strategy();
```
Comment (inoas) [inline src/ORM/Association.php]: It seems to also act as a setter and thus the docblock above should educate about that.

**Review by inoas — CHANGES_REQUESTED** [review]

File: src/ORM/Association/ExternalAssociationTrait.php [hunk 2]
```diff
@@ -1,139 +0,0 @@
...
-     * @param array $opts original list of options passed in constructor
-     * @return void
-     */
-    protected function _options(array $opts)
-    {
-        if (isset($opts['sort'])) {
-            $this->sort($opts['sort']);
-        }
-    }
-}
```
Comment (dakota) [inline src/ORM/Association/ExternalAssociationTrait.php]: Chance that anybody is using this trait?

Comment (lorenzo) [inline src/ORM/Association/ExternalAssociationTrait.php]: Very low, I have not heard of anyone creating their own associations. In any case, they have the new classes where he same methods live

File: src/ORM/Association.php (hunk 1 above)
Comment (lorenzo) [inline src/ORM/Association.php]: This is not a setter

File: src/ORM/Association.php [hunk 3]
```diff
@@ -751,6 +751,20 @@ public function deleteAll($conditions)
...
+
+        return $strategy === $this::STRATEGY_SELECT;
```
Comment (markstory) [inline src/ORM/Association.php]: `static::STRATEGY_SELECT`?

File: src/ORM/Association/BelongsTo.php [hunk 4]
```diff
@@ -30,8 +31,6 @@
 class BelongsTo extends Association
 {
 
-    use SelectableAssociationTrait;
```
Comment (markstory) [inline src/ORM/Association/BelongsTo.php]: :clap:

File: src/ORM/Association/ExternalAssociationTrait.php (hunk 2 above)
Comment (markstory) [inline src/ORM/Association/ExternalAssociationTrait.php]: I think changing internal implementation is ok for a minor release. I doubt anyone ever figured out how to use these traits. I know both @lorenzo and myself have become lost in the maze of methods they create a few times now.

File: src/ORM/Association/Loader/SelectLoader.php [hunk 5]
```diff
@@ -9,40 +9,115 @@
...
 use Cake\Database\Expression\TupleComparison;
 use Cake\Database\ValueBinder;
+use Cake\ORM\Association;
 use InvalidArgumentException;
+use RuntimeException;
 
 /**
- * Represents a type of association that that can be fetched using another query
+ * Implements the logic for loading an association using a SELECT query
  */
```
Comment (markstory) [inline src/ORM/Association/Loader/SelectLoader.php]: Should this class be marked as `@internal` or is it part of the 'public' API for the ORM?

File: src/ORM/Association/Loader/SelectLoader.php [hunk 6]
```diff
@@ -9,40 +9,115 @@
...
+     * @var string|array
+     */
+    protected $foreignKey;
+
+    /**
+     * The strategy to use for loading, either select or subquery
+     *
+     * @var string
+     */
+    protected $strategy = 'select';
```
Comment (markstory) [inline src/ORM/Association/Loader/SelectLoader.php]: Is there a constant we should use instead?

File: src/ORM/Association/Loader/SelectLoader.php [hunk 7]
```diff
@@ -77,20 +153,17 @@ protected function _defaultOptions()
...
 
-        $finder = isset($options['finder']) ? $options['finder'] : $this->finder();
-        list($finder, $opts) = $this->_extractFinder($finder);
         if (!isset($options['fields'])) {
             $options['fields'] = [];
         }
 
-        $fetchQuery = $this
-            ->find($finder, $opts)
+        $fetchQuery = $finder()
```
Comment (markstory) [inline src/ORM/Association/Loader/SelectLoader.php]: Kind of nice that we get to reuse the `find()` logic in `Association::find()` now.

File: src/ORM/Association/Loader/SelectWithPivotLoader.php [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+ * @since         3.4.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\ORM\Association\Loader;
+
+use InvalidArgumentException;
+use RuntimeException;
+
+/**
+ * Implements the logic for loading an association using a SELECT query and a pivot table
```
Comment (markstory) [inline src/ORM/Association/Loader/SelectWithPivotLoader.php]: Another `@internal` class?

**Review by markstory — COMMENTED** [review]
Awesome work @lorenzo :clap:

File: src/ORM/Association.php (hunk 1 above)
Comment (inoas) [inline src/ORM/Association.php]: Sorry I misread.

File: src/ORM/Association/Loader/SelectLoader.php (hunk 5 above)
Comment (lorenzo) [inline src/ORM/Association/Loader/SelectLoader.php]: I can make it as internal

**Comment by lorenzo** [comment]
@markstory I marked the classes as internal and added the other suggested changes

**Comment by markstory** [comment]
Looks good to me :+1:

**Pull request merged**